File: scratch/AddBvaTests.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

public class AddBvaTests {
    public static void main(String[] args) {
        String filePath = "../tests/aidims_collection.json";
        try {
            String content = new String(Files.readAllBytes(Paths.get(filePath)), "UTF-8");
            
            // Check if already added
            if (content.contains("BVA Valid Phone Length")) {
                System.out.println("BVA test cases already added!");
                return;
            }

            // The new BVA test cases as a JSON string fragment
            String bvaJsonFragment = ",\n" +
                "    {\n" +
                "      \"name\": \"User Management - Create User - BVA Valid Phone Length (20 chars)\",\n" +
                "      \"request\": {\n" +
                "        \"method\": \"POST\",\n" +
                "        \"header\": [\n" +
                "          {\n" +
                "            \"key\": \"Content-Type\",\n" +
                "            \"value\": \"application/json\"\n" +
                "          }\n" +
                "        ],\n" +
                "        \"body\": {\n" +
                "          \"mode\": \"raw\",\n" +
                "          \"raw\": \"{\\n    \\\"username\\\": \\\"bva_user_valid\\\",\\n    \\\"password\\\": \\\"password123\\\",\\n    \\\"fullName\\\": \\\"BVA User Valid Phone\\\",\\n    \\\"email\\\": \\\"bva_user_valid@example.com\\\",\\n    \\\"phone\\\": \\\"09876543210987654321\\\",\\n    \\\"role\\\": {\\n        \\\"roleId\\\": 2\\n    }\\n}\"\n" +
                "        },\n" +
                "        \"url\": {\n" +
                "          \"raw\": \"{{base_url}}/api/users\",\n" +
                "          \"host\": [\n" +
                "            \"{{base_url}}\"\n" +
                "          ],\n" +
                "          \"path\": [\n" +
                "            \"api\",\n" +
                "            \"users\"\n" +
                "          ]\n" +
                "        }\n" +
                "      },\n" +
                "      \"event\": [\n" +
                "        {\n" +
                "          \"listen\": \"test\",\n" +
                "          \"script\": {\n" +
                "            \"exec\": [\n" +
                "              \"pm.test(\\\"Status code is 200\\\", function () {\",\n" +
                "              \"    pm.response.to.have.status(200);\",\n" +
                "              \"});\",\n" +
                "              \"pm.test(\\\"User created with valid boundary phone length\\\", function () {\",\n" +
                "              \"    var jsonData = pm.response.json();\",\n" +
                "              \"    pm.expect(jsonData.status).to.eql(\\\"success\\\");\",\n" +
                "              \"    pm.expect(jsonData.data.userId).to.not.be.null;\",\n" +
                "              \"    pm.environment.set(\\\"bvaValidUserId\\\", jsonData.data.userId);\",\n" +
                "              \"});\"\n" +
                "            ],\n" +
                "            \"type\": \"text/javascript\"\n" +
                "          }\n" +
                "        }\n" +
                "      ]\n" +
                "    },\n" +
                "    {\n" +
                "      \"name\": \"User Management - Create User - BVA Invalid Phone Length (21 chars)\",\n" +
                "      \"request\": {\n" +
                "        \"method\": \"POST\",\n" +
                "        \"header\": [\n" +
                "          {\n" +
                "            \"key\": \"Content-Type\",\n" +
                "            \"value\": \"application/json\"\n" +
                "          }\n" +
                "        ],\n" +
                "        \"body\": {\n" +
                "          \"mode\": \"raw\",\n" +
                "          \"raw\": \"{\\n    \\\"username\\\": \\\"bva_user_invalid\\\",\\n    \\\"password\\\": \\\"password123\\\",\\n    \\\"fullName\\\": \\\"BVA User Invalid Phone\\\",\\n    \\\"email\\\": \\\"bva_user_invalid@example.com\\\",\\n    \\\"phone\\\": \\\"098765432109876543210\\\",\\n    \\\"role\\\": {\\n        \\\"roleId\\\": 2\\n    }\\n}\"\n" +
                "        },\n" +
                "        \"url\": {\n" +
                "          \"raw\": \"{{base_url}}/api/users\",\n" +
                "          \"host\": [\n" +
                "            \"{{base_url}}\"\n" +
                "          ],\n" +
                "          \"path\": [\n" +
                "            \"api\",\n" +
                "            \"users\"\n" +
                "          ]\n" +
                "        }\n" +
                "      },\n" +
                "      \"event\": [\n" +
                "        {\n" +
                "          \"listen\": \"test\",\n" +
                "          \"script\": {\n" +
                "            \"exec\": [\n" +
                "              \"pm.test(\\\"Status code is 500 or 400\\\", function () {\",\n" +
                "              \"    pm.expect(pm.response.code).to.be.oneOf([400, 500]);\",\n" +
                "              \"});\",\n" +
                "              \"pm.test(\\\"Error message indicates database size constraint violation\\\", function () {\",\n" +
                "              \"    var jsonData = pm.response.json();\",\n" +
                "              \"    pm.expect(jsonData.status).to.eql(\\\"error\\\");\",\n" +
                "              \"    pm.expect(jsonData.message).to.include(\\\"lỗi\\\");\",\n" +
                "              \"});\"\n" +
                "            ],\n" +
                "            \"type\": \"text/javascript\"\n" +
                "          }\n" +
                "        }\n" +
                "      ]\n" +
                "    },\n" +
                "    {\n" +
                "      \"name\": \"Doctor - Diagnostic Report - Create Report - BVA Valid Report Code Length (20 chars)\",\n" +
                "      \"request\": {\n" +
                "        \"method\": \"POST\",\n" +
                "        \"header\": [\n" +
                "          {\n" +
                "            \"key\": \"Content-Type\",\n" +
                "            \"value\": \"application/json\"\n" +
                "          }\n" +
                "        ],\n" +
                "        \"body\": {\n" +
                "          \"mode\": \"raw\",\n" +
                "          \"raw\": \"{\\n    \\\"resultId\\\": 1,\\n    \\\"reportCode\\\": \\\"BC202506240012345678\\\",\\n    \\\"findings\\\": \\\"Phổi sáng bình thường.\\\",\\n    \\\"impression\\\": \\\"Bình thường.\\\",\\n    \\\"recommendations\\\": \\\"Khám định kỳ.\\\",\\n    \\\"reportType\\\": \\\"ChinhThuc\\\",\\n    \\\"radiologistId\\\": 4,\\n    \\\"status\\\": \\\"HoanThanh\\\",\\n    \\\"referringDoctorName\\\": \\\"BS. Nguyễn Văn A\\\",\\n    \\\"referringDoctorSpecialty\\\": \\\"Tim mạch\\\"\\n}\"\n" +
                "        },\n" +
                "        \"url\": {\n" +
                "          \"raw\": \"{{base_url}}/api/diagnostic-reports\",\n" +
                "          \"host\": [\n" +
                "            \"{{base_url}}\"\n" +
                "          ],\n" +
                "          \"path\": [\n" +
                "            \"api\",\n" +
                "            \"diagnostic-reports\"\n" +
                "          ]\n" +
                "        }\n" +
                "      },\n" +
                "      \"event\": [\n" +
                "        {\n" +
                "          \"listen\": \"test\",\n" +
                "          \"script\": {\n" +
                "            \"exec\": [\n" +
                "              \"pm.test(\\\"Status code is 201\\\", function () {\",\n" +
                "              \"    pm.response.to.have.status(201);\",\n" +
                "              \"});\",\n" +
                "              \"pm.test(\\\"Diagnostic report created with valid boundary reportCode length\\\", function () {\",\n" +
                "              \"    var jsonData = pm.response.json();\",\n" +
                "              \"    pm.expect(jsonData.success).to.eql(true);\",\n" +
                "              \"    pm.expect(jsonData.data.reportId).to.not.be.null;\",\n" +
                "              \"    pm.environment.set(\\\"bvaValidReportId\\\", jsonData.data.reportId);\",\n" +
                "              \"});\"\n" +
                "            ],\n" +
                "            \"type\": \"text/javascript\"\n" +
                "          }\n" +
                "        }\n" +
                "      ]\n" +
                "    },\n" +
                "    {\n" +
                "      \"name\": \"Doctor - Diagnostic Report - Create Report - BVA Invalid Report Code Length (21 chars)\",\n" +
                "      \"request\": {\n" +
                "        \"method\": \"POST\",\n" +
                "        \"header\": [\n" +
                "          {\n" +
                "            \"key\": \"Content-Type\",\n" +
                "            \"value\": \"application/json\"\n" +
                "          }\n" +
                "        ],\n" +
                "        \"body\": {\n" +
                "          \"mode\": \"raw\",\n" +
                "          \"raw\": \"{\\n    \\\"resultId\\\": 1,\\n    \\\"reportCode\\\": \\\"BC2025062400123456789\\\",\\n    \\\"findings\\\": \\\"Phổi sáng bình thường.\\\",\\n    \\\"impression\\\": \\\"Bình thường.\\\",\\n    \\\"recommendations\\\": \\\"Khám định kỳ.\\\",\\n    \\\"reportType\\\": \\\"ChinhThuc\\\",\\n    \\\"radiologistId\\\": 4,\\n    \\\"status\\\": \\\"HoanThanh\\\",\\n    \\\"referringDoctorName\\\": \\\"BS. Nguyễn Văn A\\\",\\n    \\\"referringDoctorSpecialty\\\": \\\"Tim mạch\\\"\\n}\"\n" +
                "        },\n" +
                "        \"url\": {\n" +
                "          \"raw\": \"{{base_url}}/api/diagnostic-reports\",\n" +
                "          \"host\": [\n" +
                "            \"{{base_url}}\"\n" +
                "          ],\n" +
                "          \"path\": [\n" +
                "            \"api\",\n" +
                "            \"diagnostic-reports\"\n" +
                "          ]\n" +
                "        }\n" +
                "      },\n" +
                "      \"event\": [\n" +
                "        {\n" +
                "          \"listen\": \"test\",\n" +
                "          \"script\": {\n" +
                "            \"exec\": [\n" +
                "              \"pm.test(\\\"Status code is 400 or 500\\\", function () {\",\n" +
                "              \"    pm.expect(pm.response.code).to.be.oneOf([400, 500]);\",\n" +
                "              \"});\",\n" +
                "              \"pm.test(\\\"Error response indicates validation error or database violation\\\", function () {\",\n" +
                "              \"    var jsonData = pm.response.json();\",\n" +
                "              \"    pm.expect(jsonData.success).to.eql(false);\",\n" +
                "              \"    pm.expect(jsonData.message).to.include(\\\"Lỗi\\\");\",\n" +
                "              \"});\"\n" +
                "            ],\n" +
                "            \"type\": \"text/javascript\"\n" +
                "          }\n" +
                "        }\n" +
                "      ]\n" +
                "    },\n" +
                "    {\n" +
                "      \"name\": \"User Management - Delete BVA Valid User\",\n" +
                "      \"request\": {\n" +
                "        \"method\": \"DELETE\",\n" +
                "        \"header\": [],\n" +
                "        \"url\": {\n" +
                "          \"raw\": \"{{base_url}}/api/users/{{bvaValidUserId}}\",\n" +
                "          \"host\": [\n" +
                "            \"{{base_url}}\"\n" +
                "          ],\n" +
                "          \"path\": [\n" +
                "            \"api\",\n" +
                "            \"users\",\n" +
                "            \"{{bvaValidUserId}}\"\n" +
                "          ]\n" +
                "        }\n" +
                "      },\n" +
                "      \"event\": [\n" +
                "        {\n" +
                "          \"listen\": \"test\",\n" +
                "          \"script\": {\n" +
                "            \"exec\": [\n" +
                "              \"pm.test(\\\"Status code is 200\\\", function () {\",\n" +
                "              \"    pm.response.to.have.status(200);\",\n" +
                "              \"});\"\n" +
                "            ],\n" +
                "            \"type\": \"text/javascript\"\n" +
                "          }\n" +
                "        }\n" +
                "      ]\n" +
                "    }";

            // Find the last occurrence of ] before the final closing }
            int lastBracketIndex = content.lastIndexOf("]");
            if (lastBracketIndex == -1) {
                System.err.println("Could not find closing bracket in collection JSON!");
                return;
            }

            // Insert before the last bracket
            String newContent = content.substring(0, lastBracketIndex) + bvaJsonFragment + "\n  " + content.substring(lastBracketIndex);
            
            // Write it back
            Files.write(Paths.get(filePath), newContent.getBytes("UTF-8"));
            System.out.println("✅ BVA Test Cases successfully appended to the Postman collection via Java!");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: scratch/add_bva_tests.js
const fs = require('fs');
const path = require('path');

const collectionPath = path.join(__dirname, '../tests/aidims_collection.json');

if (!fs.existsSync(collectionPath)) {
    console.error(`Collection file not found at ${collectionPath}`);
    process.exit(1);
}

const data = JSON.parse(fs.readFileSync(collectionPath, 'utf8'));

// Define BVA test cases
const bvaTestCases = [
  {
    "name": "User Management - Create User - BVA Valid Phone Length (20 chars)",
    "request": {
      "method": "POST",
      "header": [
        {
          "key": "Content-Type",
          "value": "application/json"
        }
      ],
      "body": {
        "mode": "raw",
        "raw": JSON.stringify({
          "username": "bva_user_valid",
          "password": "password123",
          "fullName": "BVA User Valid Phone",
          "email": "bva_user_valid@example.com",
          "phone": "09876543210987654321", // Exactly 20 characters
          "role": {
            "roleId": 2
          }
        }, null, 4)
      },
      "url": {
        "raw": "{{base_url}}/api/users",
        "host": [
          "{{base_url}}"
        ],
        "path": [
          "api",
          "users"
        ]
      }
    },
    "event": [
      {
        "listen": "test",
        "script": {
          "exec": [
            "pm.test(\"Status code is 200\", function () {",
            "    pm.response.to.have.status(200);",
            "});",
            "pm.test(\"User created with valid boundary phone length\", function () {",
            "    var jsonData = pm.response.json();",
            "    pm.expect(jsonData.status).to.eql(\"success\");",
            "    pm.expect(jsonData.data.userId).to.not.be.null;",
            "    // Clean up created user ID for deletion if needed",
            "    pm.environment.set(\"bvaValidUserId\", jsonData.data.userId);",
            "});"
          ],
          "type": "text/javascript"
        }
      }
    ]
  },
  {
    "name": "User Management - Create User - BVA Invalid Phone Length (21 chars)",
    "request": {
      "method": "POST",
      "header": [
        {
          "key": "Content-Type",
          "value": "application/json"
        }
      ],
      "body": {
        "mode": "raw",
        "raw": JSON.stringify({
          "username": "bva_user_invalid",
          "password": "password123",
          "fullName": "BVA User Invalid Phone",
          "email": "bva_user_invalid@example.com",
          "phone": "098765432109876543210", // Exactly 21 characters (exceeds 20)
          "role": {
            "roleId": 2
          }
        }, null, 4)
      },
      "url": {
        "raw": "{{base_url}}/api/users",
        "host": [
          "{{base_url}}"
        ],
        "path": [
          "api",
          "users"
        ]
      }
    },
    "event": [
      {
        "listen": "test",
        "script": {
          "exec": [
            "pm.test(\"Status code is 500 or 400\", function () {",
            "    pm.expect(pm.response.code).to.be.oneOf([400, 500]);",
            "});",
            "pm.test(\"Error message indicates database size constraint violation\", function () {",
            "    var jsonData = pm.response.json();",
            "    pm.expect(jsonData.status).to.eql(\"error\");",
            "    pm.expect(jsonData.message).to.include(\"lỗi\");",
            "});"
          ],
          "type": "text/javascript"
        }
      }
    ]
  },
  {
    "name": "Doctor - Diagnostic Report - Create Report - BVA Valid Report Code Length (20 chars)",
    "request": {
      "method": "POST",
      "header": [
        {
          "key": "Content-Type",
          "value": "application/json"
        }
      ],
      "body": {
        "mode": "raw",
        "raw": "{\n    \"resultId\": 1,\n    \"reportCode\": \"{{bvaValidReportCode}}\",\n    \"findings\": \"Phổi sáng bình thường.\",\n    \"impression\": \"Bình thường.\",\n    \"recommendations\": \"Khám định kỳ.\",\n    \"reportType\": \"ChinhThuc\",\n    \"radiologistId\": 4,\n    \"status\": \"HoanThanh\",\n    \"referringDoctorName\": \"BS. Nguyễn Văn A\",\n    \"referringDoctorSpecialty\": \"Tim mạch\"\n}"
      },
      "url": {
        "raw": "{{base_url}}/api/diagnostic-reports",
        "host": [
          "{{base_url}}"
        ],
        "path": [
          "api",
          "diagnostic-reports"
        ]
      }
    },
    "event": [
      {
        "listen": "prerequest",
        "script": {
          "exec": [
            "var rand = Math.floor(Math.random() * 100000).toString().padStart(5, '0');",
            "var code = 'BC' + Date.now().toString() + rand;",
            "pm.environment.set('bvaValidReportCode', code);"
          ],
          "type": "text/javascript"
        }
      },
      {
        "listen": "test",
        "script": {
          "exec": [
            "pm.test(\"Status code is 201\", function () {",
            "    pm.response.to.have.status(201);",
            "});",
            "pm.test(\"Diagnostic report created with valid boundary reportCode length\", function () {",
            "    var jsonData = pm.response.json();",
            "    pm.expect(jsonData.success).to.eql(true);",
            "    pm.expect(jsonData.data.reportId).to.not.be.null;",
            "    pm.environment.set(\"bvaValidReportId\", jsonData.data.reportId);",
            "});"
          ],
          "type": "text/javascript"
        }
      }
    ]
  },
  {
    "name": "Doctor - Diagnostic Report - Create Report - BVA Invalid Report Code Length (21 chars)",
    "request": {
      "method": "POST",
      "header": [
        {
          "key": "Content-Type",
          "value": "application/json"
        }
      ],
      "body": {
        "mode": "raw",
        "raw": "{\n    \"resultId\": 1,\n    \"reportCode\": \"{{bvaInvalidReportCode}}\",\n    \"findings\": \"Phổi sáng bình thường.\",\n    \"impression\": \"Bình thường.\",\n    \"recommendations\": \"Khám định kỳ.\",\n    \"reportType\": \"ChinhThuc\",\n    \"radiologistId\": 4,\n    \"status\": \"HoanThanh\",\n    \"referringDoctorName\": \"BS. Nguyễn Văn A\",\n    \"referringDoctorSpecialty\": \"Tim mạch\"\n}"
      },
      "url": {
        "raw": "{{base_url}}/api/diagnostic-reports",
        "host": [
          "{{base_url}}"
        ],
        "path": [
          "api",
          "diagnostic-reports"
        ]
      }
    },
    "event": [
      {
        "listen": "prerequest",
        "script": {
          "exec": [
            "var rand = Math.floor(Math.random() * 1000000).toString().padStart(6, '0');",
            "var code = 'BC' + Date.now().toString() + rand;",
            "pm.environment.set('bvaInvalidReportCode', code);"
          ],
          "type": "text/javascript"
        }
      },
      {
        "listen": "test",
        "script": {
          "exec": [
            "pm.test(\"Status code is 400 or 500\", function () {",
            "    pm.expect(pm.response.code).to.be.oneOf([400, 500]);",
            "});",
            "pm.test(\"Error response indicates validation error or database violation\", function () {",
            "    var jsonData = pm.response.json();",
            "    pm.expect(jsonData.success).to.eql(false);",
            "    pm.expect(jsonData.message).to.include(\"Lỗi\");",
            "});"
          ],
          "type": "text/javascript"
        }
      }
    ]
  },
  // Clean up BVA created data
  {
    "name": "User Management - Delete BVA Valid User",
    "request": {
      "method": "DELETE",
      "header": [],
      "url": {
        "raw": "{{base_url}}/api/users/{{bvaValidUserId}}",
        "host": [
          "{{base_url}}"
        ],
        "path": [
          "api",
          "users",
          "{{bvaValidUserId}}"
        ]
      }
    },
    "event": [
      {
        "listen": "test",
        "script": {
          "exec": [
            "pm.test(\"Status code is 200\", function () {",
            "    pm.response.to.have.status(200);",
            "});"
          ],
          "type": "text/javascript"
        }
      }
    ]
  }
];

// Append BVA test cases
data.item.push(...bvaTestCases);

// Write back updated collection file
fs.writeFileSync(collectionPath, JSON.stringify(data, null, 2), 'utf8');
console.log('✅ BVA Test Cases successfully appended to the Postman collection!');
